core/client/src/call_executor.rs

```diff
@@ -210,64 +219,72 @@ where
 		NC: FnOnce() -> result::Result<R, &'static str> + UnwindSafe,
 	>(
 		&self,
+		initialize_block: IB,
 		at: &BlockId<Block>,
 		method: &str,
 		call_data: &[u8],
-		changes: &mut OverlayedChanges,
-		initialized_block: &mut Option<BlockId<Block>>,
-		prepare_environment_block: PB,
+		changes: &RefCell<OverlayedChanges>,
+		initialized_block: &RefCell<Option<BlockId<Block>>>,
 		execution_manager: ExecutionManager<EM>,
 		native_call: Option<NC>,
-		mut side_effects_handler: Option<&mut O>,
+		side_effects_handler: Option<&mut O>,
+		skip_initialize_block: bool,
+		recorder: &Option<Rc<RefCell<ProofRecorder<Block>>>>,
 	) -> Result<NativeOrEncoded<R>, error::Error> where ExecutionManager<EM>: Clone {
+		if !skip_initialize_block
+			&& initialized_block.borrow().as_ref().map(|id| id != at).unwrap_or(true) {
```

Comment:
Would it be worth it to coalesce `skip_initialize_block` and `initialized_block` to a enum? Not sure, but I wonder wdyt